src/model/TaskManager.cpp

```diff
@@ -71,57 +78,74 @@ TaskActionResult TaskManager::EditTask(const TaskId& id, const Task& task) {
 }
 
 TaskActionResult TaskManager::DeleteTask(const TaskId& id, bool force_delete_subtasks) {
-    auto deleting_task = tasks_.find(id);
-    if (deleting_task == tasks_.end())
+    std::lock_guard<std::mutex> guard(tasks_map_mutex_);
+
+    std::vector<TaskId> task_ids_to_delete;
+    auto task_to_delete = tasks_.find(id);
+
+    if (task_to_delete == tasks_.end())
         return TaskActionResult::FAIL_NO_SUCH_TASK;
+    else
+        task_ids_to_delete.push_back(task_to_delete->first);
 
-    auto search_child_rule = [&id](const std::pair<TaskId, TaskNode>& task) {
+    if (!force_delete_subtasks && std::find_if(tasks_.begin(), tasks_.end(), [&id](const std::pair<TaskId, TaskNode>& task) {
         return task.second.parent_id() == id;
-    };
-    if (force_delete_subtasks) {
-        auto subtask = std::find_if(tasks_.begin(), tasks_.end(), search_child_rule);
-        while (subtask != tasks_.end()) {
-            this->DeleteTask(subtask->first, true);
-            subtask = std::find_if(tasks_.begin(), tasks_.end(), search_child_rule);
-        }
-    } else if (std::find_if(tasks_.begin(), tasks_.end(), search_child_rule) != tasks_.end()) {
+    }) != tasks_.end()) {
         return TaskActionResult::FAIL_NOT_DELETED_SUBTASKS;
     }
 
-    deleted_tasks_.insert_or_assign(deleting_task->first, deleting_task->second);
-    tasks_.erase(deleting_task);
+    for (size_t i = 0; i < task_ids_to_delete.size(); ++i) {
+        for (const auto& task : tasks_) {
```

Comment:
this for-loop can be a standalone loop since task_ids_to_delete can only be of size 1 ?
and maybe it can be refactored to work with the previous FAIL_NOT_DELETED_SUBTASKS check.
Possibly same comment for Complete()